python/ray/ml/trainer.py

```diff
@@ -171,7 +188,32 @@ def preprocess_datasets(self) -> None:
         The transformed datasets will be set back in the ``self.datasets`` attribute
         of the Trainer to be used when overriding ``training_loop``.
         """
-        raise NotImplementedError
+        # Evaluate all datasets.
+        self.datasets = {k: d() if callable(d) else d for k, d in self.datasets.items()}
+
+        # Transform all datasets concurrently in remote tasks.
+        transform_task_dict = {}
+
+        transform_task = ray.remote(
+            lambda preprocessor, dataset: preprocessor.transform(dataset)
+        )
+
+        if self.preprocessor and not self._has_preprocessed_datasets:
+            train_dataset = self.datasets.get(TRAIN_DATASET_KEY, None)
```

Comment:
Based on conversation on the other PR, do we need default value?
```suggestion
            train_dataset = self.datasets[TRAIN_DATASET_KEY]
```

Reply:
Are you talking about the xgboost PR? 

I think the plan was to make a train dataset required for xgboost or lightgbm trainer, but not necessarily for DataParallelTrainer.

Reply:
Oh, what does it mean if there is no train dataset for `DataParallelTrainer`?

Reply:
DataParallelTrainer is out of scope for this PR? Like this is only the general Trainer. We should not enforce that a training dataset is required for the general trainer case.

Reply:
Oh... this question was because you mentioned "but not necessarily for DataParallelTrainer", so I just wanted to know the scenario you were referring to.

My preference would be to make it required until there is a definitive use-case that doesn't need it, but this isn't a strong/blocking opinion.